File: thinking-categorically/semigroup/csv- transform-example/csv- transform-example.cs
using System;
using System.Collections.Generic;
using System.Linq;

// Semigroup interface
public interface ISemigroup<T>
{
    T Combine(T a, T b);
}

// Person class for type safety
public class Person
{
    public string Name { get; set; }
    public string Age { get; set; }
    public int? TotalAge { get; set; }
    public int? Count { get; set; }

    public Person(string name, string age, int? totalAge = null, int? count = null)
    {
        Name = name;
        Age = age;
        TotalAge = totalAge;
        Count = count;
    }
}

// Semigroup for combining person records
public class PersonSemigroup : ISemigroup<Person>
{
    public Person Combine(Person p1, Person p2)
    {
        return new Person(
            name: $"{p1.Name} & {p2.Name}",  // Combine names
            age: Math.Max(int.Parse(p1.Age), int.Parse(p2.Age)).ToString(), // Take max age
            totalAge: (p1.TotalAge ?? int.Parse(p1.Age)) + (p2.TotalAge ?? int.Parse(p2.Age)),
            count: (p1.Count ?? 1) + (p2.Count ?? 1)
        );
    }
}

// Semigroup for aggregating ages (sum)
public class AgeSumSemigroup : ISemigroup<int>
{
    public int Combine(int a, int b) => a + b;
}

// Semigroup for collecting names
public class NameListSemigroup : ISemigroup<List<string>>
{
    public List<string> Combine(List<string> a, List<string> b)
    {
        var result = new List<string>(a);
        result.AddRange(b);
        return result;
    }
}

// Generic array semigroup
public class ArraySemigroup<T> : ISemigroup<List<T>>
{
    public List<T> Combine(List<T> a, List<T> b)
    {
        var result = new List<T>(a);
        result.AddRange(b);
        return result;
    }
}

// Statistics class for aggregating data
public class Stats
{
    public int TotalAge { get; set; }
    public int Count { get; set; }
    public List<string> Names { get; set; }

    public Stats(int totalAge, int count, List<string> names)
    {
        TotalAge = totalAge;
        Count = count;
        Names = names ?? new List<string>();
    }
}

// Semigroup for combining statistics
public class StatsSemigroup : ISemigroup<Stats>
{
    public Stats Combine(Stats s1, Stats s2)
    {
        var combinedNames = new List<string>(s1.Names);
        combinedNames.AddRange(s2.Names);

        return new Stats(
            totalAge: s1.TotalAge + s2.TotalAge,
            count: s1.Count + s2.Count,
            names: combinedNames
        );
    }
}

// Extension methods for semigroup operations
public static class SemigroupExtensions
{
    // Generic combine function for collections
    public static T CombineAll<T>(this ISemigroup<T> semigroup, IEnumerable<T> items)
    {
        var list = items.ToList();
        if (!list.Any())
        {
            throw new ArgumentException("Cannot combine empty collection");
        }

        return list.Aggregate(semigroup.Combine);
    }
}

// CSV parsing utility
public static class CsvParser
{
    public static List<Person> ReadCSV(string data)
    {
        var lines = data.Trim().Split('\n');
        if (lines.Length == 0) return new List<Person>();

        var header = lines[0];
        var rows = lines.Skip(1);
        var keys = header.Split(',');

        return rows.Select(row =>
        {
            var values = row.Split(',');
            var obj = keys.Zip(values, (k, v) => new { Key = k, Value = v })
                         .ToDictionary(x => x.Key, x => x.Value);

            return new Person(
                name: obj["name"],
                age: obj["age"]
            );
        }).ToList();
    }
}

// Example usage
class Program
{
    static void Main()
    {
        // Initialize semigroups
        var personSemigroup = new PersonSemigroup();
        var ageSumSemigroup = new AgeSumSemigroup();
        var nameListSemigroup = new NameListSemigroup();
        var arraySemigroup = new ArraySemigroup<Person>();
        var statsSemigroup = new StatsSemigroup();

        // Parse CSV data
        var data = CsvParser.ReadCSV("name,age\nAlice,25\nBob,40\nCharlie,35");

        Console.WriteLine("Original data:");
        foreach (var person in data)
        {
            Console.WriteLine($"  {person.Name}, {person.Age}");
        }

        // 1. Combine all persons into one record
        var combinedPerson = personSemigroup.CombineAll(data);
        Console.WriteLine($"\nCombined person: {combinedPerson.Name}, " +
                         $"age: {combinedPerson.Age}, total age: {combinedPerson.TotalAge}, " +
                         $"count: {combinedPerson.Count}");

        // 2. Sum all ages using semigroup
        var ages = data.Select(person => int.Parse(person.Age));
        var totalAge = ageSumSemigroup.CombineAll(ages);
        Console.WriteLine($"Total age: {totalAge}");

        // 3. Collect all names using semigroup
        var nameArrays = data.Select(person => new List<string> { person.Name });
        var allNames = nameListSemigroup.CombineAll(nameArrays);
        Console.WriteLine($"All names: [{string.Join(", ", allNames)}]");

        // 4. Process multiple CSV datasets and combine them
        var data2 = CsvParser.ReadCSV("name,age\nDave,28\nEve,32");
        var data3 = CsvParser.ReadCSV("name,age\nFrank,45\nGrace,29");

        var allDatasets = arraySemigroup.CombineAll(new[] { data, data2, data3 });
        Console.WriteLine($"\nCombined datasets ({allDatasets.Count} people):");
        foreach (var person in allDatasets)
        {
            Console.WriteLine($"  {person.Name}, {person.Age}");
        }

        // 5. Create summary statistics using semigroups
        var statsFromData = allDatasets.Select(p => new Stats(
            totalAge: int.Parse(p.Age),
            count: 1,
            names: new List<string> { p.Name }
        ));

        var overallStats = statsSemigroup.CombineAll(statsFromData);
        Console.WriteLine($"\nOverall statistics:");
        Console.WriteLine($"  Total age: {overallStats.TotalAge}");
        Console.WriteLine($"  Count: {overallStats.Count}");
        Console.WriteLine($"  Names: [{string.Join(", ", overallStats.Names)}]");
        Console.WriteLine($"  Average age: {(double)overallStats.TotalAge / overallStats.Count:F2}");
    }
}

File: thinking-categorically/semigroup/csv- transform-example/csv- transform-example.hs
import Data.List.Split (splitOn)
import Data.Semigroup (Sum(..), getSum)

-- Person data type
data Person = Person
  { name :: String
  , age :: Int
  , totalAge :: Maybe Int
  , count :: Maybe Int
  } deriving (Show, Eq)

-- Semigroup instance for Person
instance Semigroup Person where
  p1 <> p2 = Person
    { name = name p1 ++ " & " ++ name p2
    , age = max (age p1) (age p2)
    , totalAge = Just $ (maybe (age p1) id (totalAge p1)) + (maybe (age p2) id (totalAge p2))
    , count = Just $ (maybe 1 id (count p1)) + (maybe 1 id (count p2))
    }

-- Stats data type for summary statistics
data Stats = Stats
  { statsTotalAge :: Int
  , statsCount :: Int
  , statsNames :: [String]
  } deriving (Show, Eq)

-- Semigroup instance for Stats
instance Semigroup Stats where
  s1 <> s2 = Stats
    { statsTotalAge = statsTotalAge s1 + statsTotalAge s2
    , statsCount = statsCount s1 + statsCount s2
    , statsNames = statsNames s1 ++ statsNames s2
    }

-- CSV parsing function
readCSV :: String -> [Person]
readCSV csvData =
  let allLines = lines csvData
  in case allLines of
       [] -> []
       (header:rows) ->
         let keys = splitOn "," header
             parseRow row =
               let values = splitOn "," row
                   rowMap = zip keys values
                   n = lookup "name" rowMap
                   a = lookup "age" rowMap >>= readMaybe
               in case (n, a) of
                    (Just nameVal, Just ageVal) ->
                      Just $ Person nameVal ageVal Nothing Nothing
                    _ -> Nothing
         in map (\r -> case parseRow r of
                         Just p -> p
                         Nothing -> error "Malformed row") rows
  where
    readMaybe :: String -> Maybe Int
    readMaybe s = case reads s of
                    [(x, "")] -> Just x
                    _ -> Nothing

-- Combine all elements using semigroup
combineAll :: Semigroup a => [a] -> Maybe a
combineAll [] = Nothing
combineAll (x:xs) = Just $ foldl (<>) x xs

-- Convert Person to Stats
personToStats :: Person -> Stats
personToStats p = Stats (age p) 1 [name p]

-- Main function
main :: IO ()
main = do
  let csvData = "name,age\nAlice,25\nBob,40\nCharlie,35"
  let dataList = readCSV csvData

  putStrLn "Original data:"
  mapM_ print dataList

  -- 1. Combine all persons into one record
  case combineAll dataList of
    Just combinedPerson -> do
      putStrLn "\nCombined person:"
      print combinedPerson
    Nothing -> putStrLn "No data to combine"

  -- 2. Sum all ages using built-in Sum semigroup
  let ages = map age dataList
  let ageSums = map (Sum . fromIntegral :: Int -> Sum Integer) ages  -- Convert to Sum semigroup
  case combineAll ageSums of
    Just totalSum -> putStrLn $ "\nTotal age (using Sum semigroup): " ++ show (getSum totalSum)
    Nothing -> putStrLn "No ages to sum"

  -- 3. Collect all names using semigroup
  let names = map ((: []) . name) dataList  -- Convert to list of lists
  case combineAll names of
    Just allNames -> putStrLn $ "All names: " ++ show allNames
    Nothing -> putStrLn "No names to collect"

  -- 4. Process multiple CSV datasets and combine them
  let csvData2 = "name,age\nDave,28\nEve,32"
  let csvData3 = "name,age\nFrank,45\nGrace,29"
  let data2 = readCSV csvData2
  let data3 = readCSV csvData3

  let allDatasets = dataList ++ data2 ++ data3
  putStrLn "\nCombined datasets:"
  mapM_ print allDatasets

  -- 5. Create summary statistics using semigroups
  let statsFromData = map personToStats allDatasets
  case combineAll statsFromData of
    Just overallStats -> do
      putStrLn "\nOverall statistics:"
      print overallStats
      let avgAge = fromIntegral (statsTotalAge overallStats) / fromIntegral (statsCount overallStats) :: Double
      putStrLn $ "Average age: " ++ show avgAge

      -- Also demonstrate Sum semigroup for total age calculation
      let allAges = map (Sum . fromIntegral . age :: Person -> Sum Integer) allDatasets
      case combineAll allAges of
        Just totalAgeSum -> putStrLn $ "Total age (Sum semigroup): " ++ show (getSum totalAgeSum)
        Nothing -> putStrLn "No ages to sum with Sum semigroup"
    Nothing -> putStrLn "No stats to combine"
